### DISCUSSION TITLE
OpenTelemetry: How can I count requests for each downstream?

### DISCUSSION THREAD

**Discussion by tuanhd9173** [dotnet/yarp#2266] [Q&A] [answered]
Hello,

I followed the sample in order to expose forwarder metrics to Prometheus. 
But I want to expose metrics (RequestsStarted, RequestsFailed, CurrentRequests) for each downstream (each cluster) too. 
I found that I can register an IForwarderTelemetryConsumer to consume forward events but I can not identify the downstream.
Do somebody have experience with?

Thanks.

**Comment by MihaZupan** [comment, marked as answer]
The [ReverseProxy.Metrics.Sample](https://github.com/microsoft/reverse-proxy/tree/main/samples/ReverseProxy.Metrics.Sample) project shows how you can correlate those `IForwarderTelemetryConsumer` (and other) events back to the route/cluster/destination.

E.g. in [`ForwarderTelemetryConsumer.OnForwarderStart`](https://github.com/microsoft/reverse-proxy/blob/ea355627c0cbf8a0ead57bec3b9f56530f1f05f1/samples/ReverseProxy.Metrics.Sample/ForwarderTelemetryConsumer.cs#L12), you can look at `metrics.DestinationId`.

.NET 8 will include more HttpClient metrics out of the box, so there should be less of a need to expose some of these metrics manually.
